File: android-app/app/src/main/kotlin/com/subhandler/utils/NotificationUtils.kt
package com.subhandler.utils

import com.subhandler.model.Subscription

fun buildNotificationTitle(subscription: Subscription, daysUntil: Long): String = when {
    daysUntil <= 0 -> "${subscription.name} renews today"
    daysUntil == 1L -> "${subscription.name} renews tomorrow"
    else -> "${subscription.name} renews in $daysUntil days"
}

fun buildNotificationBody(subscription: Subscription, daysUntil: Long): String {
    val cost = formatSubscriptionCost(subscription)
    return when {
        daysUntil <= 0 -> "${subscription.name} renews today — $cost"
        daysUntil == 1L -> "${subscription.name} renews tomorrow — $cost"
        else -> "${subscription.name} renews in $daysUntil days — $cost"
    }
}

File: src/utils/notificationUtils.ts
import type { Subscription } from '../types';
import { formatCurrency } from './costUtils';

export function getNotificationKey(sub: Subscription): string {
  return `${sub.id}-${sub.nextRenewalDate}`;
}

export function buildNotificationPayload(sub: Subscription, daysUntil: number): { title: string; body: string } {
  const cost = formatCurrency(sub.cost, sub.currency);
  if (daysUntil <= 0) {
    return {
      title: `${sub.name} renews today`,
      body: `${sub.name} renews today — ${cost}`,
    };
  }
  if (daysUntil === 1) {
    return {
      title: `${sub.name} renews tomorrow`,
      body: `${sub.name} renews tomorrow — ${cost}`,
    };
  }
  return {
    title: `${sub.name} renews in ${daysUntil} days`,
    body: `${sub.name} renews in ${daysUntil} days — ${cost}`,
  };
}
